bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Workspace.java

```diff
@@ -2160,8 +2160,12 @@ public Resource newResource(IPath path, int type) {
 	 * modifications to the tree.
 	 */
 	public ElementTree newWorkingTree() {
-		tree = tree.newEmptyDelta();
-		return tree;
+		// synchronized for atomic swap. Should have already synchronized by
+		// getWorkManager().checkIn/checkout, but it's not guaranteed
```

Comment:
According to the comment, we might have **two** locks now involved? This is dangerous.

Reply:
newEmptyDelta() does not lock anything beside its individual synchronize.